README
NAME
    Dancer::Session::Redis - Redis backend for Dancer Session Engine

VERSION
    version 0.20

SYNOPSIS
        # in the Dancer config.yml:
        session: 'Redis'
        redis_session:
            sock: '/var/run/redis.sock'
            password: 'QmG_kZECJAvAcDaWqqSqoNLUka5v3unMe_8sqYMh6ST'
            database: 1
            expire: 3600
            debug: 0
            ping: 5

        # or in the Dancer application:
        setting redis_session => {
            server   => 'redi.example.com:6379',
            password => 'QmG_kZECJAvAcDaWqqSqoNLUka5v3unMe_8sqYMh6ST',
            database => 1,
            expire   => 3600,
            debug    => 0,
            ping     => 5,
        };
        setting session => 'Redis';

DESCRIPTION
    This module is a Redis backend for the session engine of Dancer
    application. This module is a descendant of Dancer::Session::Abstract. A
    simple demo apllication might be found in the "eg/" directory of this
    distribution.

CONFIGURATION
    In order to use this session engine, you have to set up a few settings
    (in the app or app's configuration file).

    *   session

        Set the vaue Redis. Required parameter.

    *   redis_session

        Settings for backend.

        *server*
            Hostname and port of the redis-server instance which will be
            used to store session data. This one is required unless *sock*
            is defined.

        *sock*
            unix socket path of the redis-server instance which will be used
            to store session data.

        *password*
            Password string for redis-server's AUTH command to processing
            any other commands. Optional. Check the redis-server manual for
            directive *requirepass* if you would to use redis internal
            authentication.

        *database*
            Database # to store session data. Optional. Default value is 0.

        *expire*
            Session TTL. Optional. Default value is 900 (seconds).

        *ping*
            Time (in seconds) to check connection alive and re-establish in
            case of closed connection. Optional. Default value is 5
            (seconds). Redis server close connection after a client is idle
            for seconds but server instance might be configured to not close
            client's connection. Check the redis server manual.

        *debug*
            Enables debug information to STDERR, including all interactions
            with the redis-server. Optional. Default value is 0.

METHODS
  init()
    Validate settings and creates the initial connection to redis-server.

  create()
    Creates a new object, runs "flush" and returns the object.

  flush()
    Writes the session information to the Redis database.

  retrieve()
    Retrieves session information from the Redis database.

  destroy()
    Deletes session information from the Redis database.

BUGS
    Please report any bugs or feature requests through the web interface at
    <https://github.com/Wu-Wu/Dancer-Session-Redis/issues>

SEE ALSO
    Dancer

    Dancer::Session

    Storable

    Redis

    redis.io <http://redis.io>

AUTHOR
    Anton Gerasimov, <chim@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2012 by Anton Gerasimov

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Dancer::Session::Redis.

{{$NEXT}}

0.22    Sun Apr 7 11:20:18 2013 +0000
    - start using Test::Exception in *.t
    - set minimum version of perl to 5.8.9

0.21    Sun Mar 24 17:06:21 2013 +0000
    - refactored connection establishing to redis server
    - updated and added some *.t
    - dzillify distribution

0.20  Thu Jan 17 16:37:18 2013
    - added example of usage
    - added support unix socket configurations (GH#2) [thanks to c0decafe!]
    - added license and changelog files
    - other minor fixes

0.1.9  Fri Dec 21 07:42:20 2012
    - added support for authenticated sessions

0.1.8  Fri Dec 7 12:31:19 2012
	- original version


eg/app.pl
#
# This example is a part of the Dancer-Session-Redis distribution
#

# 1) Start this application server
#       perl app.pl
# 2) Paste and go to the following link in your favourite browser
#       http://localhost:4100/
# 3) ...
# 4) PROFIT!

use strict;
use warnings;
use Dancer;

# some settings
set port          => 4100;
set logger        => 'console';

# set up session
set redis_session => {
    server   => 'localhost:6379',  # this option should be tuned to fit your needs
    database => 3,
    expire   => 3600,
    debug    => 0,
    ping     => 5,
};
set session       => 'Redis';
set session_name  => 'eg_session_id'; # session's cookie name

# route
get '/' => sub {

    # fetch old value
    my $old = session->{counter};

    # prepare new value
    my $new = 1 + (defined $old ? $old : 0);

    # save new value
    session counter => $new;

    # display result
    join '' => 'Session ID: ', session->id, '<br/>', 'Counter: ', $new, '<br/><br/>Reload the page...';

};

dance;


Makefile.PL
use 5.010;
use ExtUtils::MakeMaker;

my $meta_merge = {
    META_MERGE => {
        resources => {
            repository => 'https://github.com/Wu-Wu/Dancer-Session-Redis',
            bugtracker => 'https://github.com/Wu-Wu/Dancer-Session-Redis/issues',
        },
    }
};

my %options = (
    'NAME'          => 'Dancer::Session::Redis',
    'VERSION_FROM'  => 'lib/Dancer/Session/Redis.pm',
    'AUTHOR'        => 'Anton Gerasimov <chim@cpan.org>',
    'ABSTRACT_FROM' => 'lib/Dancer/Session/Redis.pm',
    'LICENSE'       => 'perl',
    'PREREQ_PM'     => {
        'Dancer'   => '1.3072',
        'Redis'    => '1.955',
        'strict'   => 0,
        'warnings' => 0,
        'parent'   => 0,
        'Storable' => 0,
        'Carp'     => 0,
    },
    'BUILD_REQUIRES' => {
        'Test::More' => '0.98',
        'Test::Pod'  => '1.22',
    },
    'MIN_PERL_VERSION' => '5.10.0',
    'test' => {
        'TESTS' => 't/*.t',
    },
    $ExtUtils::MakeMaker::VERSION >= 6.50 ? (%$meta_merge) : (),
);

WriteMakefile(%options);


lib/Dancer/Session/Redis.pm
package Dancer::Session::Redis;

# ABSTRACT: Redis backend for Dancer Session Engine

use strict;
use warnings;
use parent 'Dancer::Session::Abstract';
use Redis 1.955;
use Dancer::Config 'setting';
use Storable ();
use Carp ();

# VERSION
# AUTHORITY

my $_redis;
my %options = ();

sub init {
    my $self = shift;

    $self->SUPER::init(@_);

    # backend settings
    if (my $opts = setting('redis_session') ) {
        if (ref $opts and ref $opts eq 'HASH' ) {
            %options = (
                server   => $opts->{server}   || undef,
                sock     => $opts->{sock}     || undef,
                database => $opts->{database} || 0,
                expire   => $opts->{expire}   || 900,
                debug    => $opts->{debug}    || 0,
                password => $opts->{password} || undef,
            );
        }
        else {
            Carp::croak 'Settings redis_session must be a hash reference!';
        }
    }
    else {
        Carp::croak 'Settings redis_session is not defined!';
    }

    unless (defined $options{server} || defined $options{sock}) {
        Carp::croak 'Settings redis_session should include either server or sock parameter!';
    }

    # get radis handle
    $self->redis;
}

# create a new session
sub create {
    my ($class) = @_;

    $class->new->flush;
}

# fetch the session object by id
sub retrieve($$) {
    my ($class, $id) = @_;

    my $self = $class->new;
    $self->redis->select($options{database});
    $self->redis->expire($id => $options{expire});

    Storable::thaw($self->redis->get($id));
}

# delete session
sub destroy {
    my ($self) = @_;

    $self->redis->select($options{database});
    $self->redis->del($self->id);
}

# flush session
sub flush {
    my ($self) = @_;

    $self->redis->select($options{database});
    $self->redis->set($self->id => Storable::freeze($self));
    $self->redis->expire($self->id => $options{expire});

    $self;
}

# get redis handle
sub redis {
    my ($self) = @_;

    if (!$_redis || !$_redis->ping) {
        my %params = (
            debug     => $options{debug},
            reconnect => 10,
            every     => 100,
        );

        if (defined $options{sock}) {
            $params{sock} = $options{sock};
        }
        else {
            $params{server} = $options{server};
        }

        $params{password} = $options{password} if $options{password};

        $_redis = Redis->new(%params);
    }

    $_redis and return $_redis;

    Carp::croak "Unable connect to redis-server...";
}

1; # End of Dancer::Session::Redis

__END__

=pod

=head1 SYNOPSIS

    # in the Dancer config.yml:
    session: 'Redis'
    redis_session:
        sock: '/var/run/redis.sock'
        password: 'QmG_kZECJAvAcDaWqqSqoNLUka5v3unMe_8sqYMh6ST'
        database: 1
        expire: 3600
        debug: 0

    # or in the Dancer application:
    setting redis_session => {
        server   => 'redi.example.com:6379',
        password => 'QmG_kZECJAvAcDaWqqSqoNLUka5v3unMe_8sqYMh6ST',
        database => 1,
        expire   => 3600,
        debug    => 0,
    };
    setting session => 'Redis';

=head1 DESCRIPTION

This module is a Redis backend for the session engine of Dancer application. This module is a descendant
of L<Dancer::Session::Abstract>. A simple demo apllication might be found in the C<eg/> directory of this
distribution.

=head1 CONFIGURATION

In order to use this session engine, you have to set up a few settings (in the app or app's configuration file).

=head2 session

Set the vaue B<Redis>. Required parameter.

=head2 redis_session

Settings for backend.

=head3 server

Hostname and port of the redis-server instance which will be used to store session data. This one is B<required> unless I<sock> is defined.

=head3 sock

unix socket path of the redis-server instance which will be used to store session data.

=head3 password

Password string for redis-server's AUTH command to processing any other commands. Optional. Check the redis-server
manual for directive I<requirepass> if you would to use redis internal authentication.

=head3 database

Database # to store session data. Optional. Default value is 0.

=head3 expire

Session TTL. Optional. Default value is 900 (seconds).

=head3 debug

Enables debug information to STDERR, including all interactions with the redis-server. Optional. Default value is 0.

=head1 METHODS

=head2 init

Validate settings and creates the initial connection to redis-server.

=head2 redis

Returns connection handle to the redis instance. Also establish new connection in case of C<dead> handle.

=head2 create

Creates a new object, runs C<flush> and returns the object.

=head2 flush

Writes the session information to the Redis database.

=head2 retrieve

Retrieves session information from the Redis database.

=head2 destroy

Deletes session information from the Redis database.

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Dancer-Session-Redis/issues>

=head1 SEE ALSO

L<Dancer>

L<Dancer::Session>

L<Storable>

L<Redis>

L<redis.io|http://redis.io>

=cut
